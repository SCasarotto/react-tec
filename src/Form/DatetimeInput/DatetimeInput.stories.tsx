import React from 'react';

import { Story, Meta } from '@storybook/react';

import { DatetimeInput, DatetimeInputProps } from './DatetimeInput';

export default {
  title: 'Form/DatetimeInput',
  component: DatetimeInput,
  argTypes: {},
} as Meta;

const Template: Story<DatetimeInputProps> = (args) => (
  <DatetimeInput {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const Time = Template.bind({});
Time.args = {
  timeFormat: 'h:mm aa',
  timeIntervals: 15,
  dateFormat: 'MMMM d, yyyy h:mm aa',
  showTimeSelect: true,
};

export const WithPortalButton = Template.bind({});
WithPortalButton.args = {
  timeFormat: 'h:mm aa',
  timeIntervals: 15,
  dateFormat: 'MMMM d, yyyy h:mm aa',
  showTimeSelect: true,
  withPortalButton: true,
};
