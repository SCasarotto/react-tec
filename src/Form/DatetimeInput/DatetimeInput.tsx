import React from 'react';

import DatePicker, { ReactDatePickerProps } from 'react-datepicker';

import {
  WithPortalButtonWrapper,
  WithPortalButton,
  WithPortalIcon,
} from './styledComponents';

export interface DatetimeInputProps
  // TODO: Improve Type
  extends Omit<ReactDatePickerProps<any>, 'value'> {
  //These are used to standardize to other input props
  placeholder?: string;
  // TODO: Improve Type
  value?: ReactDatePickerProps<any>['selected'];
  withPortalButton?: boolean;
}
/**
 *
 * A styled [React Datepicker](https://reactdatepicker.com/) Component. A few props have been adjusted or defaulted.
 *
 * Note: Styling for this component is not handled the same way as most React C components. In order to style everything the React DatePicker css has been internalized in the React C global styles. In the future this may be adjusted to make styling easier and more consistent with other components.
 *
 */
export const DatetimeInput = React.forwardRef<
  // TODO: Improve Type
  DatePicker<any>,
  DatetimeInputProps
>((props, ref) => {
  const {
    selected,
    value,
    placeholderText,
    placeholder,
    className = '',
    withPortalButton,
    ...rest
  } = props;

  if (withPortalButton) {
    return (
      <WithPortalButtonWrapper>
        <DatePicker
          selected={selected || value}
          placeholderText={placeholderText || placeholder}
          className={`tec-datetimeInput ${className}`}
          ref={ref}
          {...rest}
        />
        <DatePicker
          selected={selected || value}
          withPortal
          customInput={
            <WithPortalButton className='tec-datetimeInputWithPortalButton'>
              <WithPortalIcon className='tec-datetimeInputWithPortalButtonIcon' />
            </WithPortalButton>
          }
          {...rest}
          id={rest.id ? rest.id + 'withPortalButton' : rest.id}
        />
      </WithPortalButtonWrapper>
    );
  }

  return (
    <DatePicker
      selected={selected || value}
      placeholderText={placeholderText || placeholder}
      className={`tec-datetimeInput ${className}`}
      ref={ref}
      {...rest}
    />
  );
});

DatetimeInput.defaultProps = {
  autoComplete: 'off',
  // dateFormat: 'MM/DD/YYYY',
};
