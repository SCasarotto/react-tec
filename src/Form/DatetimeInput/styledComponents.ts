import { FaRegCalendarAlt } from 'react-icons/fa';
import styled from 'styled-components';

import { Button } from '../Button';

export const WithPortalButtonWrapper = styled.div`
  display: flex;

  .react-datepicker__input-container input {
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }
`;
export const WithPortalButton = styled(Button)`
  position: relative;
  display: block;
  width: auto;
  font-size: 17px;
  padding: 10px 12px;
  box-sizing: border-box;
  cursor: pointer;
  border-radius: 5px;
  border-top-left-radius: 0px;
  border-bottom-left-radius: 0px;
  background-color: ${(props) => props.theme.white};
  color: ${(props) => props.theme.darkerGray};
  text-align: left;

  border: 1px solid ${(props) => props.theme.lightGray};
  transition: border-color 0.2s ease-in, box-shadow 0.2s ease-in;

  :hover,
  :active {
    border: 1px solid ${(props) => props.theme.primary};
    box-shadow: 0 0 0 1px ${(props) => props.theme.primary} inset;
  }
  :disabled {
    color: ${(props) => props.theme.gray};
    border: 1px solid ${(props) => props.theme.lighterGray};
    box-shadow: none;
    cursor: not-allowed;
  }
`;
export const WithPortalIcon = styled(FaRegCalendarAlt)`
  color: inherit;
  display: block;
`;
