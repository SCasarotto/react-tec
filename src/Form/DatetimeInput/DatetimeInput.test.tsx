import React from 'react';

import { render, waitFor } from '@testing-library/react';
import DatePicker from 'react-datepicker';

import { DatetimeInput } from './DatetimeInput';

describe('DatetimeInput', () => {
  // Renders Without Error
  it('matches snapshot', () => {
    const { asFragment } = render(<DatetimeInput onChange={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  // Class Name
  it('accepts custom className', () => {
    const { container } = render(
      <DatetimeInput className='myClassName' onChange={jest.fn()} />,
    );
    expect(container.querySelector('.myClassName')).toBeTruthy();
  });

  // Ref
  it('renders with ref', async () => {
    const ref: { current: DatePicker | null } = { current: null };
    render(<DatetimeInput onChange={jest.fn()} ref={ref} />);
    waitFor(() => expect(ref.current).toBeTruthy());
  });
});
