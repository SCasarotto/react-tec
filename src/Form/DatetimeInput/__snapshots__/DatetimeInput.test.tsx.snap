// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatetimeInput matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        autocomplete="off"
        class="tec-datetimeInput "
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
